File: src/ui/analytics.py
"""
ui/analytics.py
================
Streamlit "Analytics" page: usage dashboard (documents, chunks,
embeddings, response time, storage, most-queried documents, upload
activity chart) plus a Settings panel for chunking, retrieval, and LLM
parameters.

Presented as two tabs on one page (Analytics / Settings) rather than
two separate top-level files — both are small enough that a dedicated
`ui/settings.py` would be more ceremony than value, and both consume
the same underlying `AnalyticsService`/`RuntimeSettingsService` pair.
"""

from __future__ import annotations

import streamlit as st

from src.services.analytics_service import AnalyticsService
from src.services.settings_service import RuntimeSettingsService
from src.utils.exceptions import DocIntelError, ValidationError


def render(analytics_service: AnalyticsService, settings_service: RuntimeSettingsService) -> None:
    """Render the full Analytics + Settings page."""
    tab_analytics, tab_settings = st.tabs(["📊 Analytics", "⚙️ Settings"])

    with tab_analytics:
        _render_analytics(analytics_service)

    with tab_settings:
        _render_settings(settings_service)


# ------------------------------------------------------------------
# Analytics dashboard
# ------------------------------------------------------------------
def _render_analytics(analytics_service: AnalyticsService) -> None:
    st.subheader("Usage Analytics")

    try:
        summary = analytics_service.get_summary()
    except DocIntelError as exc:
        st.error(f"Couldn't load analytics: {exc}")
        return

    row1 = st.columns(4)
    row1[0].metric("Total Documents", summary.total_documents)
    row1[1].metric("Total Chunks", summary.total_chunks)
    row1[2].metric("Total Embeddings", summary.total_embeddings)
    row1[3].metric("Storage Used", summary.total_storage_display)

    row2 = st.columns(4)
    row2[0].metric("Chat Sessions", summary.total_chat_sessions)
    row2[1].metric("Total Queries", summary.total_queries)
    row2[2].metric("Avg. Response Time", summary.average_response_time_display)
    row2[3].metric(
        "Most Queried",
        summary.most_queried_documents[0].filename if summary.most_queried_documents else "N/A",
    )

    st.divider()

    col_activity, col_ranking = st.columns(2)
    with col_activity:
        st.markdown("**Upload Activity (last 14 days)**")
        if summary.uploads_per_day:
            chart_data = {day: count for day, count in summary.uploads_per_day}
            st.bar_chart(chart_data)
        else:
            st.caption("No uploads yet in this period.")

    with col_ranking:
        st.markdown("**Most Queried Documents**")
        if summary.most_queried_documents:
            for entry in summary.most_queried_documents:
                st.markdown(f"**{entry.filename}**")
                st.caption(f"{entry.query_count} quer{'y' if entry.query_count == 1 else 'ies'}")
        else:
            st.caption("No queries logged yet.")


# ------------------------------------------------------------------
# Settings panel
# ------------------------------------------------------------------
def _render_settings(settings_service: RuntimeSettingsService) -> None:
    st.subheader("Settings")
    st.caption(
        "Changes to Top-K, Temperature, Max Tokens, and Model apply to your "
        "next chat message. Chunk Size/Overlap apply to documents uploaded "
        "after saving — already-processed documents are unaffected."
    )

    try:
        current = settings_service.get_preferences()
    except DocIntelError as exc:
        st.error(f"Couldn't load current settings: {exc}")
        return

    with st.form("settings_form"):
        st.markdown("**Document Chunking**")
        col1, col2 = st.columns(2)
        with col1:
            chunk_size = st.slider("Chunk Size (characters)", min_value=100, max_value=4000, value=current.chunk_size, step=50)
        with col2:
            chunk_overlap = st.slider("Chunk Overlap (characters)", min_value=0, max_value=max(chunk_size - 1, 1), value=min(current.chunk_overlap, max(chunk_size - 1, 1)), step=10)

        st.markdown("**Retrieval**")
        top_k = st.slider("Top-K Retrieval", min_value=1, max_value=20, value=current.top_k)

        st.markdown("**LLM Generation**")
        col3, col4 = st.columns(2)
        with col3:
            temperature = st.slider("Temperature", min_value=0.0, max_value=2.0, value=current.temperature, step=0.1)
        with col4:
            max_tokens = st.slider("Maximum Tokens", min_value=64, max_value=8192, value=current.max_tokens, step=64)

        model_options = RuntimeSettingsService.available_models()
        model_index = model_options.index(current.model) if current.model in model_options else 0
        model = st.selectbox("Model", options=model_options, index=model_index)

        col_save, col_reset = st.columns(2)
        with col_save:
            save_clicked = st.form_submit_button("Save Settings", type="primary", use_container_width=True)
        with col_reset:
            reset_clicked = st.form_submit_button("Reset to Defaults", use_container_width=True)

    if save_clicked:
        try:
            settings_service.update_preferences(
                chunk_size=chunk_size,
                chunk_overlap=chunk_overlap,
                top_k=top_k,
                temperature=temperature,
                max_tokens=max_tokens,
                model=model,
            )
            st.success("Settings saved.")
        except ValidationError as exc:
            st.error(f"Couldn't save settings: {exc}")

    if reset_clicked:
        try:
            settings_service.reset_to_defaults()
            st.success("Settings reset to defaults.")
            st.rerun()
        except DocIntelError as exc:
            st.error(f"Couldn't reset settings: {exc}")

File: frontend/src/pages/analytics.js
/**
 * pages/analytics.js
 * ===================
 * Full Analytics + Settings page — feature parity with
 * `src/ui/analytics.py`: usage dashboard (8 metrics, upload-activity
 * chart, most-queried ranking) and a Settings form (chunking,
 * retrieval, LLM generation parameters).
 *
 * Presented as two sub-tabs on one page, same rationale as the
 * Streamlit version: both are small, both read from the same
 * services, and a dedicated route for Settings would be more
 * ceremony than value.
 */

import { analytics as analyticsApi, settings as settingsApi } from "../api/resources.js";
import { showToast } from "../components/toast.js";
import { escapeHtml } from "../utils/format.js";

export function renderAnalyticsPage(container) {
  let activeTab = "analytics";
  let disposed = false;

  container.innerHTML = `
    <div class="page-content stack">
      <div class="page-header">
        <h1>📈 Analytics & Settings</h1>
        <p class="text-muted">Usage dashboard and retrieval/generation preferences.</p>
      </div>
      <div class="subtabs">
        <button class="subtab-btn active" data-tab="analytics">📊 Analytics</button>
        <button class="subtab-btn" data-tab="settings">⚙️ Settings</button>
      </div>
      <div id="tab-content"></div>
    </div>
  `;

  const tabButtons = container.querySelectorAll(".subtab-btn");
  const tabContent = container.querySelector("#tab-content");

  tabButtons.forEach((button) => {
    button.addEventListener("click", () => {
      if (button.dataset.tab === activeTab) return;
      activeTab = button.dataset.tab;
      tabButtons.forEach((b) => b.classList.toggle("active", b === button));
      renderActiveTab();
    });
  });

  function renderActiveTab() {
    if (disposed) return;
    if (activeTab === "analytics") renderAnalyticsTab(tabContent);
    else renderSettingsTab(tabContent);
  }

  renderActiveTab();

  return function dispose() {
    disposed = true;
  };
}

// ----------------------------------------------------------------------
// Analytics tab
// ----------------------------------------------------------------------
async function renderAnalyticsTab(content) {
  content.innerHTML = `<div class="row"><span class="spinner"></span><span>Loading analytics...</span></div>`;

  let summary;
  try {
    summary = await analyticsApi.getSummary();
  } catch (err) {
    content.innerHTML = `<div class="alert alert-error">Couldn't load analytics: ${escapeHtml(err.message)}</div>`;
    return;
  }

  const mostQueriedName = summary.most_queried_documents[0]?.filename ?? "N/A";

  content.innerHTML = `
    <div class="stack">
      <div class="grid-metrics">
        ${metricCard("Total Documents", summary.total_documents)}
        ${metricCard("Total Chunks", summary.total_chunks)}
        ${metricCard("Total Embeddings", summary.total_embeddings)}
        ${metricCard("Storage Used", summary.total_storage_display)}
      </div>
      <div class="grid-metrics">
        ${metricCard("Chat Sessions", summary.total_chat_sessions)}
        ${metricCard("Total Queries", summary.total_queries)}
        ${metricCard("Avg. Response Time", summary.average_response_time_display)}
        ${metricCard("Most Queried", mostQueriedName)}
      </div>

      <div class="row" style="align-items: stretch; flex-wrap: wrap">
        <div class="card" style="flex: 1; min-width: 280px">
          <h3>Upload Activity (last 14 days)</h3>
          <div id="upload-chart"></div>
        </div>
        <div class="card" style="flex: 1; min-width: 280px">
          <h3>Most Queried Documents</h3>
          <div id="most-queried-list"></div>
        </div>
      </div>
    </div>
  `;

  content.querySelector("#upload-chart").appendChild(renderUploadActivityChart(summary.uploads_per_day));
  content.querySelector("#most-queried-list").appendChild(
    renderMostQueriedList(summary.most_queried_documents)
  );
}

function metricCard(label, value) {
  return `
    <div class="metric">
      <div class="metric-label">${escapeHtml(label)}</div>
      <div class="metric-value">${escapeHtml(String(value))}</div>
    </div>
  `;
}

function renderMostQueriedList(entries) {
  const wrapper = document.createElement("div");
  if (entries.length === 0) {
    wrapper.innerHTML = `<p class="text-muted">No queries logged yet.</p>`;
    return wrapper;
  }
  wrapper.className = "stack";
  wrapper.innerHTML = entries
    .map(
      (entry) => `
        <div>
          <p style="margin: 0"><strong>${escapeHtml(entry.filename)}</strong></p>
          <p class="text-muted" style="margin: 0">
            ${entry.query_count} quer${entry.query_count === 1 ? "y" : "ies"}
          </p>
        </div>
      `
    )
    .join("");
  return wrapper;
}

/** Plain inline SVG bar chart — no charting library for a handful of daily counts. */
function renderUploadActivityChart(uploadsPerDay) {
  const wrapper = document.createElement("div");

  if (!uploadsPerDay || uploadsPerDay.length === 0) {
    wrapper.innerHTML = `<p class="text-muted">No uploads yet in this period.</p>`;
    return wrapper;
  }

  const width = 280;
  const height = 140;
  const padding = { top: 8, bottom: 22, left: 4, right: 4 };
  const plotWidth = width - padding.left - padding.right;
  const plotHeight = height - padding.top - padding.bottom;
  const maxCount = Math.max(...uploadsPerDay.map(([, count]) => count), 1);
  const barGap = 4;
  const barWidth = plotWidth / uploadsPerDay.length - barGap;

  const bars = uploadsPerDay
    .map(([day, count], index) => {
      const barHeight = (count / maxCount) * plotHeight;
      const x = padding.left + index * (barWidth + barGap);
      const y = padding.top + (plotHeight - barHeight);
      const label = day.slice(5); // "MM-DD"
      return `
        <rect class="bar-chart-bar" x="${x}" y="${y}" width="${barWidth}" height="${Math.max(barHeight, 1)}" rx="2">
          <title>${escapeHtml(day)}: ${count} upload${count !== 1 ? "s" : ""}</title>
        </rect>
        ${
          count > 0
            ? `<text class="bar-chart-value" x="${x + barWidth / 2}" y="${y - 3}" text-anchor="middle">${count}</text>`
            : ""
        }
        <text class="bar-chart-label" x="${x + barWidth / 2}" y="${height - 6}" text-anchor="middle">${label}</text>
      `;
    })
    .join("");

  wrapper.innerHTML = `
    <svg viewBox="0 0 ${width} ${height}" style="width: 100%; height: auto" role="img" aria-label="Uploads per day, last 14 days">
      ${bars}
    </svg>
  `;
  return wrapper;
}

// ----------------------------------------------------------------------
// Settings tab
// ----------------------------------------------------------------------
async function renderSettingsTab(content) {
  content.innerHTML = `<div class="row"><span class="spinner"></span><span>Loading settings...</span></div>`;

  let current;
  let models;
  try {
    [current, models] = await Promise.all([
      settingsApi.get(),
      settingsApi.listModels().then((r) => r.models),
    ]);
  } catch (err) {
    content.innerHTML = `<div class="alert alert-error">Couldn't load current settings: ${escapeHtml(err.message)}</div>`;
    return;
  }

  content.innerHTML = `
    <div class="card stack">
      <div>
        <p class="text-muted">
          Changes to Top-K, Temperature, Max Tokens, and Model apply to your next chat
          message. Chunk Size/Overlap apply to documents uploaded after saving —
          already-processed documents are unaffected.
        </p>
      </div>

      <div>
        <h3>Document Chunking</h3>
        <div class="settings-grid">
          ${sliderField("chunk-size", "Chunk Size (characters)", 100, 4000, 50, current.chunk_size)}
          ${sliderField("chunk-overlap", "Chunk Overlap (characters)", 0, current.chunk_size - 1, 10, current.chunk_overlap)}
        </div>
      </div>

      <div>
        <h3>Retrieval</h3>
        ${sliderField("top-k", "Top-K Retrieval", 1, 20, 1, current.top_k)}
      </div>

      <div>
        <h3>LLM Generation</h3>
        <div class="settings-grid">
          ${sliderField("temperature", "Temperature", 0, 2, 0.1, current.temperature)}
          ${sliderField("max-tokens", "Maximum Tokens", 64, 8192, 64, current.max_tokens)}
        </div>
        <div class="field">
          <label class="field-label" for="model-select">Model</label>
          <select id="model-select" class="select">
            ${models.map((m) => `<option value="${escapeHtml(m)}" ${m === current.model ? "selected" : ""}>${escapeHtml(m)}</option>`).join("")}
          </select>
        </div>
      </div>

      <div class="row">
        <button class="btn btn-primary" id="save-settings-btn" style="flex: 1">Save Settings</button>
        <button class="btn" id="reset-settings-btn" style="flex: 1">Reset to Defaults</button>
      </div>
      <div id="settings-error"></div>
    </div>
  `;

  wireSlider(content, "chunk-size");
  wireSlider(content, "chunk-overlap");
  wireSlider(content, "top-k");
  wireSlider(content, "temperature");
  wireSlider(content, "max-tokens");

  // Chunk overlap's valid range depends on chunk size — re-clamp live,
  // mirroring the Streamlit slider's dynamic `max_value=chunk_size - 1`.
  const chunkSizeInput = content.querySelector("#chunk-size-input");
  const chunkOverlapInput = content.querySelector("#chunk-overlap-input");
  chunkSizeInput.addEventListener("input", () => {
    const newMax = Math.max(Number(chunkSizeInput.value) - 1, 1);
    chunkOverlapInput.max = newMax;
    if (Number(chunkOverlapInput.value) > newMax) {
      chunkOverlapInput.value = newMax;
      chunkOverlapInput.dispatchEvent(new Event("input"));
    }
  });

  const errorBox = content.querySelector("#settings-error");
  const saveBtn = content.querySelector("#save-settings-btn");
  const resetBtn = content.querySelector("#reset-settings-btn");

  saveBtn.addEventListener("click", async () => {
    saveBtn.disabled = true;
    errorBox.innerHTML = "";
    try {
      await settingsApi.update({
        chunk_size: Number(content.querySelector("#chunk-size-input").value),
        chunk_overlap: Number(content.querySelector("#chunk-overlap-input").value),
        top_k: Number(content.querySelector("#top-k-input").value),
        temperature: Number(content.querySelector("#temperature-input").value),
        max_tokens: Number(content.querySelector("#max-tokens-input").value),
        model: content.querySelector("#model-select").value,
      });
      showToast("Settings saved.");
    } catch (err) {
      errorBox.innerHTML = `<div class="alert alert-error">Couldn't save settings: ${escapeHtml(err.message)}</div>`;
    } finally {
      saveBtn.disabled = false;
    }
  });

  resetBtn.addEventListener("click", async () => {
    resetBtn.disabled = true;
    try {
      await settingsApi.reset();
      showToast("Settings reset to defaults.");
      await renderSettingsTab(content); // reload with fresh defaults
    } catch (err) {
      errorBox.innerHTML = `<div class="alert alert-error">Couldn't reset settings: ${escapeHtml(err.message)}</div>`;
      resetBtn.disabled = false;
    }
  });
}

function sliderField(id, label, min, max, step, value) {
  return `
    <div class="slider-field">
      <div class="slider-field-header">
        <label for="${id}-input">${escapeHtml(label)}</label>
        <span class="slider-field-value" id="${id}-value">${value}</span>
      </div>
      <input type="range" id="${id}-input" min="${min}" max="${max}" step="${step}" value="${value}" />
    </div>
  `;
}

function wireSlider(content, id) {
  const input = content.querySelector(`#${id}-input`);
  const valueLabel = content.querySelector(`#${id}-value`);
  input.addEventListener("input", () => {
    valueLabel.textContent = input.value;
  });
}
